cloud/amazon/s3.py

```diff
@@ -590,13 +608,21 @@ def main():
 
     # Support for listing a set of keys
     if mode == 'list':
-        bucket_object = get_bucket(module, s3, bucket)
-
-        # If the bucket does not exist then bail out
-        if bucket_object is None:
-            module.fail_json(msg="Target bucket (%s) cannot be found"% bucket, failed=True)
-
-        list_keys(module, bucket_object, prefix, marker, max_keys)
+        if bucket == '/':
+            try:
+                all_buckets = s3.get_all_buckets()
+                s3_keys = [x.name for x in all_buckets]
+            except s3.provider.storage_response_error, e:
```

Comment:
Need to make sure that these compile under python3 (we'll shoot for actually working under python3 later).  It looks like boto is a requirement here and boto requires python>=2.6 so we can just do:
``` python
except s3.provider.storage_response_error as e:
```

catching exceptions in files that do not have libraries which require python >= 2.6 would use something like this instead:
``` python
from ansible.module_utils.py24compat import get_exception
[...]
except s3.provider.storage_response_error:
    e = get_exception()
```